File: string/1525_number_of_good_ways_to_split_a_string/1525_number_of_good_ways_to_split_a_string.py
from collections import Counter


class Solution:
    def numSplits(self, s: str) -> int:
        left = Counter()
        right = Counter(s)
        count = 0
        for i in range(len(s)):
            left[s[i]] += 1
            right[s[i]] -= 1
            if right[s[i]] == 0:
                del right[s[i]]
                
            if len(left) == len(right):
                count += 1
        return count


s = "aacaba"
res = Solution().numSplits(s)
print(res)

File: string/1525_number_of_good_ways_to_split_a_string/1525_number_of_good_ways_to_split_a_string.go
package main

import "fmt"

func numSplits(s string) int {
	freqLeft := [26]int{}
	freqRight := [26]int{}
	for i := 0; i < len(s); i++ {
		freqRight[s[i]-'a']++
	}
	var res int
	for i := 0; i < len(s); i++ {
		freqLeft[s[i]-'a']++
		freqRight[s[i]-'a']--
		if freq(freqLeft) == freq(freqRight) {
			res++
		}
	}
	return res
}

func freq(fre [26]int) int {
	var res int
	for i := 0; i < len(fre); i++ {
		if fre[i] > 0 {
			res++
		}
	}
	return res
}

func main() {
	s := "aacaba"
	res := numSplits(s)
	fmt.Println(res)
}
